packages/client/src/publish/MessageFactory.ts

```diff
@@ -0,0 +1,105 @@
+import { random } from 'lodash'
+import { EncryptedGroupKey, EncryptionType, EthereumAddress, StreamID, StreamMessage, StreamPartID, toStreamPartID } from 'streamr-client-protocol'
+import { CacheConfig } from '../Config'
+import { EncryptionUtil } from '../encryption/EncryptionUtil'
+import { GroupKeyId } from '../encryption/GroupKey'
+import { CacheFn } from '../utils/caches'
+import { MessageChain } from './MessageChain'
+import { MessageMetadata } from './Publisher'
+import { keyToArrayIndex } from '@streamr/utils'
+import { GroupKeySequence } from './GroupKeyQueue'
+
+export interface MessageFactoryOptions {
+    streamId: StreamID
+    partitionCount: number
+    isPublicStream: boolean
+    publisherId: EthereumAddress
+    createSignature: (payload: string) => Promise<string>
+    useGroupKey: () => Promise<GroupKeySequence>
+    cacheConfig?: CacheConfig
+}
+
+export class MessageFactory {
+
+    private readonly streamId: StreamID
+    private readonly partitionCount: number
+    private readonly selectedDefaultPartition: number
+    private readonly isPublicStream: boolean
+    private readonly publisherId: EthereumAddress
+    private readonly createSignature: (payload: string) => Promise<string>
+    private readonly useGroupKey: () => Promise<GroupKeySequence>
+    private readonly getStreamPartitionForKey: (partitionKey: string | number) => number
+    private readonly getMsgChain: (streamPartId: StreamPartID, publisherId: EthereumAddress, msgChainId?: string) => MessageChain
+
+    constructor(opts: MessageFactoryOptions) {
+        this.streamId = opts.streamId
+        this.partitionCount = opts.partitionCount
+        this.selectedDefaultPartition = random(opts.partitionCount - 1)
+        this.isPublicStream = opts.isPublicStream
+        this.publisherId = opts.publisherId
+        this.createSignature = opts.createSignature
+        this.useGroupKey = opts.useGroupKey
+        this.getStreamPartitionForKey = CacheFn((partitionKey: string | number) => {
+            return keyToArrayIndex(opts.partitionCount, partitionKey)
+        }, {
+            ...opts.cacheConfig,
+            cacheKey: ([partitionKey]) => partitionKey
+        })
+        this.getMsgChain = CacheFn((streamPartId: StreamPartID, publisherId: EthereumAddress, msgChainId?: string) => { // TODO would it ok to just use pMemoize (we don't have many chains)
+            return new MessageChain(streamPartId, publisherId, msgChainId)
+        }, {
+            cacheKey: ([streamPartId, publisherId, msgChainId]) => [streamPartId, publisherId, msgChainId ?? ''].join('|'),
+            ...opts.cacheConfig,
+            maxAge: Infinity
+        })
+    }
+
+    async createMessage<T>(
```

Comment:
Inside this method we need to add the check that whether current user is valid publisher to this stream.